src/realm/object-store/impl/object_notifier.cpp

```diff
@@ -23,42 +23,97 @@
 using namespace realm;
 using namespace realm::_impl;
 
-ObjectNotifier::ObjectNotifier(std::shared_ptr<Realm> realm, TableKey table, ObjKey obj)
+ObjectNotifier::ObjectNotifier(std::shared_ptr<Realm> realm, TableKey table_key, ObjKey obj_key)
     : CollectionNotifier(std::move(realm))
-    , m_table(table)
-    , m_obj(obj)
+    , m_table_key(table_key)
+    , m_obj_key(obj_key)
 {
 }
 
+void ObjectNotifier::do_attach_to(Transaction& sg)
+{
+    m_table = sg.get_table(m_table_key);
+}
+
 bool ObjectNotifier::do_add_required_change_info(TransactionChangeInfo& info)
 {
+    if (!m_table_key)
+        return false;
+
     m_info = &info;
-    info.tables[m_table.value];
-    return false;
+    info.tables[m_table_key.value];
+
+    // When adding or removing a callback the related tables can change due to the way we calculate related tables
+    // when key path filters are set hence we need to recalculate every time the callbacks are changed.
+    util::CheckedLockGuard lock(m_callback_mutex);
+    if (m_did_modify_callbacks) {
+        update_related_tables(*m_table);
```

Comment:
I'd suggest using REALM_ASSERT(m_table) right before using m_table, and same for below when you use it in `run()`

Reply:
Added with 3296929.